Compiled flex bias appears to not work at all

### 🐛 Describe the bug

```python
import torch

from torch.nn.attention.flex_attention import flex_attention

N_BATCH = 16
N_HEAD = 32
N_CTX = 4096
D_QK = 128

torch.set_default_device("cuda")

@torch.compile
def main():
    q = torch.randn(N_BATCH, N_HEAD, N_CTX, D_QK, requires_grad=True)
    k = torch.randn(N_BATCH, N_HEAD, N_CTX, D_QK, requires_grad=True)
    v = torch.randn(N_BATCH, N_HEAD, N_CTX, D_QK, requires_grad=True)

    # source: https://github.com/pytorch/pytorch/blob/53fc921ce2bcfd29b0adc42b72b86a982a690e30/test/inductor/test_flex_attention.py#L2978
    #bias_1 = torch.randn(N_BATCH, N_HEAD, N_CTX, N_CTX, requires_grad=True)
    #def score_mod(score, b, h, q_idx, kv_idx):
    #    return score + bias_1[b, h, q_idx, kv_idx]
    # illegal memory access

    # source: https://github.com/pytorch/pytorch/blob/53fc921ce2bcfd29b0adc42b72b86a982a690e30/test/inductor/test_flex_attention.py#L4714
    #bias_2 = torch.randn(N_CTX, N_CTX, requires_grad=True)
    #def score_mod(score, b, h, q_idx, kv_idx):
    #    return score + bias_2[q_idx, kv_idx]
    # AssertionError: size=[16, 32, 4096, 128], stride=[4096, 1]

    # source: https://github.com/pytorch/pytorch/blob/53fc921ce2bcfd29b0adc42b72b86a982a690e30/test/inductor/test_flex_attention.py#L209
    #bias_3 = torch.randn(N_HEAD, requires_grad=True)
    #def score_mod(score, b, h, q_idx, kv_idx):
    #    return score * bias_3[h]
    # AssertionError: size=[16, 32, 4096, 128], stride=[1]

    flex_attention(q, k, v, score_mod=score_mod).sum().backward()
main()
```

I have taken three examples of bias from the flex tests, so I expect them to work. The first one gets an illegal memory access. The second and third get deep assertion errors. This is on latest nightly.

### Versions

```
Collecting environment information...
PyTorch version: 2.7.0.dev20250128+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov  6 2024, 20:22:13) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-49-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA H100 PCIe
Nvidia driver version: 550.127.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               26
On-line CPU(s) list:                  0-25
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8480+
CPU family:                           6
Model:                                143
Thread(s) per core:                   1
Core(s) per socket:                   1
Socket(s):                            26
Stepping:                             8
BogoMIPS:                             4000.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 wbnoinvd arat vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b fsrm md_clear serialize tsxldtrk avx512_fp16 arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            832 KiB (26 instances)
L1i cache:                            832 KiB (26 instances)
L2 cache:                             104 MiB (26 instances)
L3 cache:                             416 MiB (26 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-25
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Unknown: No mitigations
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] flake8==4.0.1
[pip3] numpy==1.21.5
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] optree==0.13.1
[pip3] pytorch-triton==3.2.0+gitb2684bf3
[pip3] torch==2.7.0.dev20250128+cu124
[pip3] triton==3.1.0
[conda] Could not collect
```

There's two issues here:

1. Dynamo seems to handle the cases where we're creating leaf tensors within the compiled region incorrectly. cc: @anijain2305 That leads to the bugs you're seeing with gradient size mismatch.
2. We're not handling int64 indexing appropriately in general right now with FlexAttention. With a 16 x 32 x 4096 x 4096 tensor, this is just enough to overflow and cause the IMA.

If you modify it to work around these issues then they go away. Still bugs, but the feature's not completely broken :P 

```
import torch

from torch.nn.attention.flex_attention import flex_attention

N_BATCH = 16
N_HEAD = 32
N_CTX = 2048
D_QK = 128

torch.set_default_device("cuda")

def main():
    q = torch.randn(N_BATCH, N_HEAD, N_CTX, D_QK, requires_grad=True)
    k = torch.randn(N_BATCH, N_HEAD, N_CTX, D_QK, requires_grad=True)
    v = torch.randn(N_BATCH, N_HEAD, N_CTX, D_QK, requires_grad=True)

    # source: https://github.com/pytorch/pytorch/blob/53fc921ce2bcfd29b0adc42b72b86a982a690e30/test/inductor/test_flex_attention.py#L2978
    # bias_1 = torch.randn(N_BATCH, N_HEAD, N_CTX, N_CTX, requires_grad=True)
    # def score_mod(score, b, h, q_idx, kv_idx):
    #    return score + bias_1[b, h, q_idx, kv_idx]
    # illegal memory access

    # source: https://github.com/pytorch/pytorch/blob/53fc921ce2bcfd29b0adc42b72b86a982a690e30/test/inductor/test_flex_attention.py#L4714
    # bias_2 = torch.randn(N_CTX, N_CTX, requires_grad=True)
    # def score_mod(score, b, h, q_idx, kv_idx):
    #    return score + bias_2[q_idx, kv_idx]
    # AssertionError: size=[16, 32, 4096, 128], stride=[4096, 1]

    # source: https://github.com/pytorch/pytorch/blob/53fc921ce2bcfd29b0adc42b72b86a982a690e30/test/inductor/test_flex_attention.py#L209
    bias_3 = torch.randn(N_HEAD, requires_grad=True)
    def score_mod(score, b, h, q_idx, kv_idx):
       return score * bias_3[h]
    # AssertionError: size=[16, 32, 4096, 128], stride=[1]

    torch.compile(flex_attention)(q, k, v, score_mod=score_mod).sum().backward()
main()
```

--- a/torch/nn/attention/flex_attention.py
+++ b/torch/nn/attention/flex_attention.py
@@ -970,7 +970,13 @@ def _nested_mod_func_adapter(
     if is_score_mod:
 
         def nt_score_mod(score, b, h, q_idx, kv_idx):
-            b_nested = q_seq_idx[q_idx]
+            
+            # Convert g_idx and kv_idx to int32
+            q_idx = q_idx.to(torch.int32)
+            kv_idx = kv_idx.to(torch.int32)
+           
+            # Convert q_idx to int if it has one element, otherwise use it as is
+            b_nested = q_seq_idx[q_idx.item()] if q_idx.numel() == 1 else q_seq_idx[q_idx]
             q_nested = q_idx - q_offsets[q_seq_idx[q_idx]]
             kv_nested = kv_idx - kv_offsets[kv_seq_idx[kv_idx]]
             is_same_sequence = q_seq_idx[q_idx] == kv_seq_idx[kv_idx]
@@ -985,7 +991,13 @@ def _nested_mod_func_adapter(
     else:
 
         def nt_mask_mod(b, h, q_idx, kv_idx):
-            b_nested = q_seq_idx[q_idx]
+
+            # Convert g_idx and kv_idx to int32
+            q_idx = q_idx.to(torch.int32)
+            kv_idx = kv_idx.to(torch.int32)
+            
+            # Convert q_idx to int if it has one element, otherwise use it as is
+            b_nested = q_seq_idx[q_idx.item()] if q_idx.numel() == 1 else q_seq_idx[q_idx]
             q_nested = q_idx - q_offsets[q_seq_idx[q_idx]]
             kv_nested = kv_idx - kv_offsets[kv_seq_idx[kv_idx]]
             # don't allow inter-sequence attention

